### PR TITLE
Add Tests for BIP-340 test vectors

### LINKED ISSUE #13: Add Test Vectors from BIP-340
Using the CSV from [BIP-340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki), create an integration test that verifies proper handling of the test vectors.

Direct links:

* Test vectors: [test-vectors.csv](https://github.com/bitcoin/bips/blob/master/bip-0340/test-vectors.csv)
* Python reference implementation: [reference.py](https://github.com/bitcoin/bips/blob/master/bip-0340/reference.py)

PR #305 (or equivalent)  is needed to use the `aux_rand` values from the test vectors.

### PR DESCRIPTION
**Pull request by liamgilligan**
Bitcoin core devs have made a [CSV](https://github.com/bitcoin/bips/blob/master/bip-0340/test-vectors.csv) of 19 test vectors. They test edge cases for Schnorr signing, signature verification, and public key generation. This PR adds tests for the aforementioned procedures using their CSV.

NOTE: There are two minor caveats to this. First, `secp256k1-jdk` does NOT support messages of length != 32 bytes. We simply skip test vectors with messages != 32 bytes when needed. Second, the test vectors have cases to test if Schnorr signature verification will reject invalid public keys (e.g. off-curve public keys). In this implementation, it isn't possible to pass an invalid public key to `schnorrSigVerify`, so instead there is a separate test for checking if `ecPubKeyParse` recognizes the given public keys as invalid.

### CODE REVIEW

**Review by msgilligan — CHANGES_REQUESTED** [review]
A made a quick first-pass review and have a few minor requested changes.

Nice work!

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/Secp256k1ForeignTest.java [hunk 1]
```diff
@@ -48,4 +54,82 @@ void pubKeyAdditionTestTwo() {
...
+    @CsvFileSource(resources = "/general-test-vectors.csv", useHeadersInDisplayName = true)
+    void pubKeyGenFromPrivKeyTestVectors(int index, @Nullable String privKeyString, String pubKeyString, String auxRand, @Nullable String messageString, String signatureString, String resultString, String comment) {
+        try (Secp256k1Foreign sec = new Secp256k1Foreign()) {
+            if (privKeyString != null) {
+                var privKey = SecpPrivKey.of(hexToBytes(privKeyString));
+                var pubKey = sec.ecPubKeyParse(
+                                ByteBuffer.allocate(33)
+                                        .put((byte) 0x02)                 // assumed even parity
+                                        .put(hexToBytes(pubKeyString))
+                                        .array())
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/Secp256k1ForeignTest.java]: This should be a helper method like `hexToBytes`.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/Secp256k1ForeignTest.java [hunk 2]
```diff
@@ -48,4 +54,82 @@ void pubKeyAdditionTestTwo() {
...
+            if (messageString != null && messageString.length() == 32) {
+                var pubKey = sec.ecPubKeyParse(
+                                ByteBuffer.allocate(33)
+                                        .put((byte) 0x02)           // even parity
+                                        .put(hexToBytes(pubKeyString))
+                                        .array())
+                        .get()
+                        .xOnly();
+                byte[] messageBytes = hexToBytes(messageString);
+                var signature = new SchnorrSignatureImpl(hexToBytes(signatureString));
```
[the remark this answers is not in the document]
Comment (msgilligan) [reply secp-integration-test/src/test/java/org/bitcoinj/secp/integration/Secp256k1ForeignTest.java]: See PR #314

**Comment by schildbach** [comment]
If I may add a tiny nitpick about the commit message. The effect of the patch is usually described in present tense (rather than past tense) because it represents something that happens the moment the patch is applied (e.g. merged to master).

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import org.jspecify.annotations.Nullable;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvFileSource;
+
+import java.util.HexFormat;
+
+
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java:80]: There should be a Javadoc comment here with links to BIP-340, the original test vectors CSV file, and maybe even the Python script.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+public class BIP340TestVectorTests {
+    @ParameterizedTest
+    @CsvFileSource(resources = "/general-test-vectors.csv", useHeadersInDisplayName = true)
+    void schnorrSigGenTestVectors(int index, @Nullable String privKeyString, String pubKeyString, String auxRand, @Nullable String messageString, String signatureString, String resultString, String comment) {
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: I would name this method `schnorrSigSign32` ~~TestVectors~~ -- since `schnorrSigSign32` is the "method-under-test". When we add a new method that supports variable-length messages, we will add a new test method to test it.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        try (Secp256k1Foreign sec = new Secp256k1Foreign()) {
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: We should use `secp` as the variable name here. It what we use in our other code (and even project like the Rust secp256k1 library use the name `secp`)

We also might be able to initialize `secp` in the constructor and thereby avoid the `try` block in each test.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            // Currently not supporting variable length messages
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: I would change this comment to say "schnorrSigSign32 only supports 32-byte messages, skip the others" or something like that.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                assertArrayEquals(expectedSignature.bytes(), actualSignature.bytes());
+            }
+        }
+    }
+
+    @ParameterizedTest
+    @CsvFileSource(resources = "/general-test-vectors.csv", useHeadersInDisplayName = true)
+    void schnorrSigVerifyTestVectors(int index, @Nullable String privKeyString, String pubKeyString, String auxRand, @Nullable String messageString, String signatureString, String resultString, String comment) {
+        try (Secp256k1Foreign sec = new Secp256k1Foreign()) {
+            // Currently not supporting variable length messages
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Doesn't verify support non-32-byte messages? Can we remove the `if`?

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            if (messageString != null && messageString.length() == 32) {
+                var pubKey = sec.xOnlyPubKeyParse(hexToBytes(pubKeyString)).get();
+                byte[] messageBytes = hexToBytes(messageString);
+                var signature = SchnorrSignature.of(hexToBytes(signatureString));
+
+                boolean expectedResult = resultString.equals("TRUE");
+                assertEquals(expectedResult, sec.schnorrSigVerify(signature, messageBytes, pubKey).get());
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: I would keep the call to the method-under-test on its own line to call extra attention to it.  eg:
```
boolean result = secp.schnorrSigVerify(signature, messageBytes, pubKey).get()

assertEquals(expectedResult, result);
```

Update: I added a more detailed suggestion about using blank lines to separate "given", "when", and "then" sections. See the overall review comment.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java (hunk 4 above)
Comment (msgilligan) [reply secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Actually, since all the test methods in this class use test vectors, we can remove `TestVectors` from all of them.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    @ParameterizedTest
+    @CsvFileSource(resources = "/general-test-vectors.csv", useHeadersInDisplayName = true)
+    void pubKeyGenFromPrivKeyTestVectors(int index, @Nullable String privKeyString, String pubKeyString, String auxRand, @Nullable String messageString, String signatureString, String resultString, String comment) {
+        try (Secp256k1Foreign sec = new Secp256k1Foreign()) {
+            if (privKeyString != null) {
+                var privKey = SecpPrivKey.of(hexToBytes(privKeyString));
+                var actualPub = sec.ecPubKeyCreate(privKey).xOnly();
+                assertEquals(pubKeyString, actualPub.toString().toUpperCase());
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: I don't think the `toUpperCase()` is needed anymore.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    void pubKeyParseFromInvalidBytesTestVectors(int index, @Nullable String privKeyString, String pubKeyString, String auxRand, @Nullable String messageString, String signatureString, String resultString, String comment) {
+        try (Secp256k1Foreign sec = new Secp256k1Foreign()) {
+            SecpResult<SecpXOnlyPubKey> result = sec.xOnlyPubKeyParse(hexToBytes(pubKeyString));
+            assertFalse(result.isOk());
+            int resultCode = (result instanceof SecpResult.Err<SecpXOnlyPubKey> err) ? err.code() : 1;
+            assertEquals(0, resultCode);
+        }
+    }
+
+    private static byte[] hexToBytes(String hex) {
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: I would rename this method to `parseHex` to match the name of the standard JDK method (and it's 2 characters and one word shorter, too!)

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            }
+        }
+    }
+
+    @ParameterizedTest
+    @CsvFileSource(resources = "/invalid-pubkey-test-vectors.csv", useHeadersInDisplayName = true)
+    void pubKeyParseFromInvalidBytesTestVectors(int index, @Nullable String privKeyString, String pubKeyString, String auxRand, @Nullable String messageString, String signatureString, String resultString, String comment) {
+        try (Secp256k1Foreign sec = new Secp256k1Foreign()) {
+            SecpResult<SecpXOnlyPubKey> result = sec.xOnlyPubKeyParse(hexToBytes(pubKeyString));
+            assertFalse(result.isOk());
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: I think we can remove this assert as it is fairly clearly redundant.

**Review by msgilligan — CHANGES_REQUESTED** [review]
Looking good! We're getting closer...

I would also suggest using the following general formatting pattern for these tests, where we divide statements into three sections:

<given-statement(s)>

<when-statement(s)>

<then-statements(s)>

Some test frameworks (e.g. Spock) use a given/when/then structure where:

given = defining all the preconditions and/or expected results  for the  tests
when = executing the code that is actually being tested
then = the assertions verifying proper behavior.

You could also look at the three sections as "setup", "execute", "verify"...

So I'm suggesting we use blank lines between each of the three sections and (generally) no blank lines within each section. I think that will help make it really easy to quickly see what each tests does.

We also had talked about using only a single CSV file that is identical to the BIP-340 CSV file. We can make that change in a separate PR.

I think most of these changes are fairy easy to make. 

Are you sure that we are testing everything the Python scripts test? The only omission that I think I spotted was that we are not testing `schnorrSigVerify()` on the non-32-byte messages.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java (hunk 11 above)
Comment (msgilligan) [reply secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: And now that PR #327 is merged lines 84-86 can be combined to a single line.

**Comment by liamgilligan** [comment]
Note about most recent push:

Added the CSV parsing @msgilligan was talking about above.

I left commits unsquashed, but will squash after this next review.

Javadoc comment needs to be written.

One more test case needs to be added for the invalid signature case.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        }
+        TestVector(String[] f) {
+            this(f[0], f[1], f[2], f[3], f[4], f[5], f[6], f[7]);
+        }
+    }
+
+    private static Stream<TestVector> schnorrSigVectors() {
+        return parseCSV()
+                .filter(vec -> vec.privKeyBytes.length == 32 && vec.messageBytes.length == 32);
+    }
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Replace this with:

static final List<TestVector> ALL_VECTORS = parseCSV();
static final List<TestVector> SIGN32_VECTORS = bip340Vectors.stream().filter(vec -> vec.privKeyBytes.length == 32 && vec.messageBytes.length == 32).toList();

and so on... defining lists for each subset of the vectors needed.

Do we need to check `vec.privKeyBytes.length == 32` or just `vec.privKeyBytes != null`?

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+            return new CSVReaderBuilder(reader)
+                    .withSkipLines(1)
+                    .build()
+                    .readAll()
+                    .stream()
+                    .map(TestVector::new);
+        } catch (IOException | CsvException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: I would put parseCSV() and any other helper method _after_ the tests. So anyone looking at the file will see the important stuff first.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        return parseCSV()
+                .filter(vec -> vec.privKeyBytes.length > 0);
+    }
+
+    private static Stream<TestVector> pubKeyParseFromInvalidBytes() {
+        return parseCSV()
+                .filter(vec -> vec.index == 5 || vec.index == 14);
+    }
+
+    private static Stream<TestVector> parseCSV() {
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Should return List<TestVector> rather than Stream.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.HexFormat;
+import java.util.stream.Stream;
+
+
+public class BIP340TestVectorTests {
+
+    private record TestVector(int index, byte[] privKeyBytes, byte[] pubKeyBytes, byte[] auxRandBytes, byte[] messageBytes, byte[] signatureBytes, boolean result, String comment){
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: I think we should drop the `Bytes` suffix on all the `byte[]` field names.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+            this(
+                    Integer.parseInt(indexString),
+                    parseHex(privKeyString),
+                    parseHex(pubKeyString),
+                    parseHex(auxRand),
+                    parseHex(messageString),
+                    parseHex(signatureString),
+                    resultString.equals("TRUE"),
+                    comment
+            );
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: For increased conciseness, consider moving this overloaded construct to a static helper method that can be placed below the tests.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java (hunk 15 above)
Comment (msgilligan) [reply secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Javadoc comments for each field in the record would be nice.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+                    .map(TestVector::new);
+        } catch (IOException | CsvException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("schnorrSigVectors")
+    void schnorrSigSign32(TestVector vec) {
+        try (Secp256k1Foreign secp = new Secp256k1Foreign()) {
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: We can move the `Secp256k1Foreign secp = new Secp256k1Foreign()` to the test class constructor and get rid of the `try` block in each test.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+            // schnorrSigSign32 only supports 32-byte messages, skip the others
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: This comment (or an improved version of it) should be in the Javadoc header for this method and/or in the Javadoc for the input `List<TestVector>`.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+                    .withSkipLines(1)
+                    .build()
+                    .readAll()
+                    .stream()
+                    .map(TestVector::new);
+        } catch (IOException | CsvException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java:78]: Example Javadoc comment for a parameterized test:
```
/// For each vector in the subset containing valid signatures on 32-byte messages, verify that the `secp` instance generates the exact, byte-for-byte signature in the [TestVector] argument.
```

**Review by msgilligan — CHANGES_REQUESTED** [review]
Getting closer. I saw your comments about remaining TODOs and agree.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java (hunk 15 above)
Comment (msgilligan) [reply secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Consider renaming `result` to `expectedVerifyResult` or something similar for clarity.

Update: or maybe `verificationResult`, which is the heading in the CSV.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java (hunk 12 above)
Comment (liamgilligan) [reply secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: It should be checking the length, it will either be length 0 or length 32.

Comment (msgilligan) [reply secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: > It should be checking the length, it will either be length 0 or length 32.

Oh, so the absence of a private key in the vector shows up as a private key of length zero? Why not just check for `length != 0` then?  (Although a `null` encoding might be better, but if that's not how the CVS library parses it, I guess it's ok to use `!= 0`)

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+                parseHex(messageString),
+                parseHex(signatureString),
+                verificationResultString.equals("TRUE"),
+                comment
+        );
+    }
+
+    private static TestVector parseVector(String[] f) {
+        return parseVector(f[0], f[1], f[2], f[3], f[4], f[5], f[6], f[7]);
+    }
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Can we combine the two parseVector() functions to eliminate the big method parameter list:
 (`String indexString, String privKeyString, String pubKeyString, String auxRandString, String messageString, String signatureString, String verificationResultString, String comment`) 
 
 To make the combined code easier to read, you could have comments on each line like:
 ```
 parseHex(f[1]),   // private key
 ```

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+        assertEquals(0, result.errorCode());
+    }
+
+    @ParameterizedTest
+    @FieldSource("INVALIDSIGNATURE_VECTORS")
+    void schnorrSigFromInvalidBytes(TestVector vec) {
+        assertThrows(IllegalArgumentException.class, () -> SchnorrSignature.of(vec.signature));
+    }
+
+    static final List<TestVector> ALL_VECTORS = parseCSV();
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Move the definition of `ALL_VECTORS` to before all the tests.

Then move the definition of each test-specific vector just before the test that tests it.

That way the comment describing what is filtered out will be right next to the comment describing the test.

**Review by msgilligan — CHANGES_REQUESTED** [review]
2 last things, besides comments.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+/**
+ * Tests public key generation from a private key and Schnorr signing and verification
+ * from a [CSV](https://github.com/bitcoin/bips/blob/master/bip-0340/test-vectors.csv)
+ * of test vectors provided in [BIP-0340](https://github.com/bitcoin/bips/tree/master/bip-0340).
+ * Based on the provided [Python reference](https://github.com/bitcoin/bips/blob/master/bip-0340/reference.py).
+ */
+public class BIP340TestVectorTests {
+
+    /**
+     * Represents a single test vector in test-vectors.csv
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Binary values represented as hex strings in the CSV are represented as byte[] in this record.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+
+/**
+ * Tests public key generation from a private key and Schnorr signing and verification
+ * from a [CSV](https://github.com/bitcoin/bips/blob/master/bip-0340/test-vectors.csv)
+ * of test vectors provided in [BIP-0340](https://github.com/bitcoin/bips/tree/master/bip-0340).
+ * Based on the provided [Python reference](https://github.com/bitcoin/bips/blob/master/bip-0340/reference.py).
+ */
+public class BIP340TestVectorTests {
+
+    /**
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Convert all `/**` Javadoc to markdown Javadoc using `///`. IntelliJ can do the conversion for you, but I'm not sure how it will handle your existing use of markdown links in these comments.

**Review by msgilligan — CHANGES_REQUESTED** [review]
Ever so close.

**Review by msgilligan — APPROVED** [review]
Approved.

File: secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+                parseHex(f[5]),             // signature
+                f[6].equals("TRUE"),        // verification result
+                f[7]                        // comment
+        );
+    }
+
+    private static byte[] parseHex(String hex) {
+        return HexFormat.of().parseHex(hex);
+    }
+
```
Comment (msgilligan) [inline secp-integration-test/src/test/java/org/bitcoinj/secp/integration/BIP340TestVectorTests.java]: Extra blank line.

**Review by msgilligan — CHANGES_REQUESTED** [review]
One extremely trivial request.

**Review by msgilligan — APPROVED** [review]
Approved! Thanks!

**Review by schildbach — APPROVED** [review]
I currently have no way to run these tests *and* inspect the results in a meaningful way, but from examination of the source code this looks great!

File: secp-integration-test/src/test/resources/test-vectors.csv [hunk 25]
```diff
@@ -0,0 +1,20 @@
+index,secret key,public key,aux_rand,message,signature,verification result,comment
```
Comment (schildbach) [inline secp-integration-test/src/test/resources/test-vectors.csv:1]: One late comment: I think this file should be moved into the same Java package that also has the consuming test class.

Comment (msgilligan) [reply secp-integration-test/src/test/resources/test-vectors.csv:1]: I agree. This is common practice with Java JUnit tests.
